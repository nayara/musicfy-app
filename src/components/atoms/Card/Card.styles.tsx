import styled, { css } from "styled-components";

type TContainerProps = {
  $clickable?: boolean;
  $imgPosition?: string;
};

const Container = styled.div<TContainerProps>`
  display: flex;
  align-items: center;
  border-radius: 0.5rem;
  padding: 1rem;

  ${({ $imgPosition }) => {
    switch ($imgPosition) {
      case "start":
        return {
          flexDirection: "row",
        };

      default:
        return {
          flexDirection: "column",
        };
    }
  }}

  ${({ $clickable }) =>
    $clickable &&
    css`
      transition: transform 0.45s;

      &:hover {
        background-color: rgba(0, 0, 0, 0.2);
      }
    `}
`;

const Button = styled.button`
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;
`;

type TTextProps = {
  $imgPosition?: string;
};

const Title = styled.h2<TTextProps>`
  text-align: left;
  color: ${({ theme }) => theme.colors.neutral.white};
  font-size: 2vh;
  align-self: flex-start;
  margin: 0;

  ${({ $imgPosition }) => {
    switch ($imgPosition) {
      case "start":
        return css`
          flex-direction: row;
        `;

      default:
        return css`
          padding: 0.3rem 0;
        `;
    }
  }}
`;

const Subtitle = styled.h3<TTextProps>`
  text-align: left;
  color: ${({ theme }) => theme.colors.terciary};
  font-size: 1.5vh;
  align-self: flex-start;
  margin: 0;
`;

type TImgProps = {
  size?: string;
};

const Image = styled.img<TImgProps>`
  border-radius: 0.5rem;

  ${({ size }) => {
    switch (size) {
      case "small":
        return css`
          height: 5vh;
        `;

      case "medium":
        return css`
          height: 20vh;
        `;

      case "large":
        return css`
          height: 40vh;
        `;

      default:
        return css`
          height: 20vh;
        `;
    }
  }}
`;

const ContentWrapper = styled.div<TContainerProps>`
  display: flex;

  ${({ $imgPosition }) => {
    switch ($imgPosition) {
      case "start":
        return css`
          height: 100%;
          flex-direction: column;
          justify-content: space-between;
          padding: 0 0.9rem;
        `;

      default:
        return css`
          flex-direction: column;
          justify-content: space-between;
          padding: 0.9rem 0;
        `;
    }
  }}
`;

export { Container, Image, Title, Button, ContentWrapper, Subtitle };
