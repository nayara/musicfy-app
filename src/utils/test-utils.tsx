import { RenderOptions, render } from "@testing-library/react";
import { FC, ReactElement, ReactNode } from "react";
import { Providers } from "src/providers";

const AllTheProviders: FC<{ children: ReactNode }> = ({ children }) => {
  return <Providers>{children}</Providers>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
