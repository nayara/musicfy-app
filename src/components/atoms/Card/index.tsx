import { FC } from "react";

import * as Style from "./Card.styles";

type TCard = {
  imgSrc: string;
  altText: string;
  title: string;
  onClick: () => void;
};

const Card: FC<TCard> = ({ imgSrc, altText, title, onClick }) => {
  return (
    <Style.Container>
      <Style.Button onClick={onClick}>
        <Style.Image src={imgSrc} alt={altText} />
        <Style.Title>{title}</Style.Title>
      </Style.Button>
    </Style.Container>
  );
};

export default Card;
