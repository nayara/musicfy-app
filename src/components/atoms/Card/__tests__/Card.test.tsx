import { userEvent } from "@testing-library/user-event";
import Card from "..";
import { render, screen } from "../../../../utils/test-utils";

describe("<Card />", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", async () => {
    const { container } = render(
      <Card
        imgSrc="some-awsome.jpg"
        altText="awsome description"
        title="Awsome image"
        onClick={() => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should call function when clicked", async () => {
    const mockedFn = jest.fn();

    render(
      <Card
        imgSrc="some-awsome.jpg"
        altText="awsome description"
        title="Awsome image"
        onClick={mockedFn}
      />
    );

    const card = screen.getByRole("button", {
      name: /awsome description/i,
    });

    await userEvent.click(card);

    expect(mockedFn).toHaveBeenCalledTimes(1);
  });
});
