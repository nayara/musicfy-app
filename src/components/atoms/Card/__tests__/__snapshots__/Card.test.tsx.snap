// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render correctly clickable card 1`] = `
.c0 {
  display: flex;
  align-items: center;
  border-radius: 0.5rem;
  padding: 1rem;
  flex-direction: column;
  transition: transform 0.45s;
}

.c0:hover {
  background-color: rgba(0, 0, 0, 0.2);
}

.c1 {
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;
}

.c4 {
  text-align: left;
  color: #EEF1EF;
  font-size: 2vh;
  align-self: flex-start;
  margin: 0;
  padding: 0.3rem 0;
}

.c2 {
  border-radius: 0.5rem;
  height: 20vh;
}

.c3 {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 0.9rem 0;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
    >
      <img
        alt="awsome description"
        class="c2"
        src="some-awsome.jpg"
      />
      <div
        class="c3"
      >
        <h2
          class="c4"
        >
          Awsome image
        </h2>
      </div>
    </button>
  </div>
</div>
`;

exports[`<Card /> should render correctly unclickable card 1`] = `
.c0 {
  display: flex;
  align-items: center;
  border-radius: 0.5rem;
  padding: 1rem;
  flex-direction: column;
}

.c3 {
  text-align: left;
  color: #EEF1EF;
  font-size: 2vh;
  align-self: flex-start;
  margin: 0;
  padding: 0.3rem 0;
}

.c1 {
  border-radius: 0.5rem;
  height: 20vh;
}

.c2 {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 0.9rem 0;
}

<div>
  <div
    class="c0"
  >
    <img
      alt="awsome description"
      class="c1"
      src="some-awsome.jpg"
    />
    <div
      class="c2"
    >
      <h2
        class="c3"
      >
        Awsome image
      </h2>
    </div>
  </div>
</div>
`;
